environment/src/test/java/com/sequenceiq/environment/environment/flow/upgrade/ccm/handler/UpgradeCcmOnFreeIpaHandlerTest.java

```diff
@@ -0,0 +1,157 @@
+
+        lenient().when(mockEnvironmentDto.getAccountId()).thenReturn(TEST_ACCOUNT_ID);
+        lenient().when(mockEnvironmentDto.getResourceCrn()).thenReturn(TEST_ENV_CRN);
+        lenient().when(mockEnvironmentDto.getName()).thenReturn(TEST_ENV_NAME);
+        lenient().when(mockEnvironmentDto.getId()).thenReturn(TEST_ENV_ID);
+        lenient().when(mockEnvironmentDto.getResourceId()).thenReturn(TEST_ENV_ID);
+        lenient().doAnswer(i -> null).when(mockEventSender).sendEvent(baseNamedFlowEvent.capture(), any(Event.Headers.class));
+    }
+
+    @Test
+    void selector() {
+        assertEquals(UPGRADE_CCM_FREEIPA_HANDLER.name(), underTest.selector());
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = Status.class, names = { "AVAILABLE", "UPGRADE_CCM_FAILED" }, mode = EnumSource.Mode.INCLUDE)
+    void testFreeIpaAcceptedStatuses(Status status) {
+        DescribeFreeIpaResponse freeipa = new DescribeFreeIpaResponse();
+        freeipa.setStatus(status);
+        freeipa.setAvailabilityStatus(AvailabilityStatus.AVAILABLE);
+        Optional<DescribeFreeIpaResponse> freeipaOpt = Optional.of(freeipa);
+        when(freeIpaService.describe(any())).thenReturn(freeipaOpt);
+        underTest.accept(mockEnvironmentDtoEvent);
+
+        verify(freeIpaPollerService).waitForCcmUpgrade(TEST_ENV_ID, TEST_ENV_CRN);
+        UpgradeCcmEvent capturedUpgradeCcmEvent = (UpgradeCcmEvent) baseNamedFlowEvent.getValue();
+        assertThat(capturedUpgradeCcmEvent.getResourceName()).isEqualTo(TEST_ENV_NAME);
+        assertThat(capturedUpgradeCcmEvent.getResourceId()).isEqualTo(TEST_ENV_ID);
+        assertThat(capturedUpgradeCcmEvent.getResourceCrn()).isEqualTo(TEST_ENV_CRN);
+        assertThat(capturedUpgradeCcmEvent.selector()).isEqualTo("UPGRADE_CCM_TUNNEL_UPDATE_EVENT");
+    }
+
+    @ParameterizedTest
+    @EnumSource(value = Status.class, names = { "AVAILABLE", "UPGRADE_CCM_FAILED" }, mode = EnumSource.Mode.EXCLUDE)
+    void testFreeIpaWrongStatuses(Status status) {
+        DescribeFreeIpaResponse freeipa = new DescribeFreeIpaResponse();
+        freeipa.setStatus(status);
+        freeipa.setAvailabilityStatus(AvailabilityStatus.AVAILABLE);
+        Optional<DescribeFreeIpaResponse> freeipaOpt = Optional.of(freeipa);
+        when(freeIpaService.describe(any())).thenReturn(freeipaOpt);
+        underTest.accept(mockEnvironmentDtoEvent);
+
+        verify(freeIpaPollerService, never()).waitForCcmUpgrade(any(), any());
+        UpgradeCcmFailedEvent capturedUpgradeCcmEvent = (UpgradeCcmFailedEvent) baseNamedFlowEvent.getValue();
+        assertThat(capturedUpgradeCcmEvent.getResourceName()).isEqualTo(TEST_ENV_NAME);
+        assertThat(capturedUpgradeCcmEvent.getResourceId()).isEqualTo(TEST_ENV_ID);
+        assertThat(capturedUpgradeCcmEvent.getResourceCrn()).isEqualTo(TEST_ENV_CRN);
+        assertThat(capturedUpgradeCcmEvent.selector()).isEqualTo("FAILED_UPGRADE_CCM_EVENT");
```

Comment:
Please also verify `EnvironmentStatus.UPGRADE_CCM_ON_FREEIPA_FAILED` if possible. Also applies to the other failure cases below.

Reply:
ok